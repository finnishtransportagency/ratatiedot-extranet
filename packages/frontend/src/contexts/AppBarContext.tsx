import React, { useEffect, useState } from 'react';
import { useLocation, useParams } from 'react-router-dom';
import { useGetUserRightPageContent } from '../hooks/query/GetUserRightPageContent';
import { getCategoryRouteName } from '../routes';

export const AppBarContext = React.createContext({
  openDrawer: false,
  toggleDrawer: () => {},
  openSearch: false,
  toggleSearch: () => {},
  openFilter: false,
  toggleFilter: () => {},
  openEdit: false,
  toggleEdit: () => {},
  openToolbar: false,
  openToolbarHandler: () => {},
  closeToolbarHandler: () => {},
  closeToolbarWithoutSaveHandler: () => {},
  userRight: { canRead: false, canWrite: false },
  userRightHandler: (_: TUserRight) => {},
});

export type TUserRight = { canRead: boolean; canWrite: boolean };

export const AppBarContextProvider = (props: any) => {
  const [openDrawer, setOpenDrawer] = useState(false);
  const [openSearch, setOpenSearch] = useState(false);
  const [openFilter, setOpenFilter] = useState(false);
  const [openEdit, setOpenEdit] = useState(false);
  const [openToolbar, setOpenToolbar] = useState(false);
  const [userRight, setUserRight] = useState<TUserRight>({ canRead: false, canWrite: false });

  const location = useLocation();
  const categoryName = getCategoryRouteName(location);
  const { area = '' } = useParams<{ area: string }>();
  const { data } = useGetUserRightPageContent(!area ? categoryName : area);

  useEffect(() => {
    if (data) {
      setUserRight(data);
    } else {
      userRightHandler({ canRead: false, canWrite: false });
      closeEdit();
    }
  }, [data]);

  const toggleDrawer = () => setOpenDrawer(!openDrawer);
  const toggleSearch = () => {
    setOpenDrawer(false);
    setOpenEdit(false);
    setOpenSearch(!openSearch);
  };
  const toggleFilter = () => {
    setOpenDrawer(false);
    setOpenEdit(false);
    setOpenFilter(!openFilter);
  };

  const toggleEdit = () => {
    setOpenDrawer(false);
    setOpenEdit(!openEdit);
  };

  const openToolbarHandler = () => {
    setOpenEdit(false);
    setOpenToolbar(true);
  };

  const closeToolbarHandler = () => {
    setOpenEdit(true);
    setOpenToolbar(false);
  };

  const closeEdit = () => {
    setOpenEdit(false);
  };

  const closeToolbarWithoutSaveHandler = () => {
    setOpenToolbar(false);
  };

  const userRightHandler = (userRight: TUserRight) => {
    setUserRight(userRight);
  };

  return (
    <AppBarContext.Provider
      value={{
        openDrawer: openDrawer,
        toggleDrawer: toggleDrawer,
        openSearch: openSearch,
        toggleSearch: toggleSearch,
        openFilter: openFilter,
        toggleFilter: toggleFilter,
        openEdit: openEdit,
        toggleEdit: toggleEdit,
        openToolbar: openToolbar,
        openToolbarHandler: openToolbarHandler,
        closeToolbarHandler: closeToolbarHandler,
        closeToolbarWithoutSaveHandler: closeToolbarWithoutSaveHandler,
        userRight: userRight,
        userRightHandler: userRightHandler,
      }}
    >
      {props.children}
    </AppBarContext.Provider>
  );
};
