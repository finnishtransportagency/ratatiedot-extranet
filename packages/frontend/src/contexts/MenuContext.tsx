import React, { useEffect, useState } from 'react';
import InfoIcon from '@mui/icons-material/Info';
import LogoutIcon from '@mui/icons-material/Logout';
import { Typography } from '@mui/material';
import SsidChartIcon from '@mui/icons-material/SsidChart';
import DirectionsIcon from '@mui/icons-material/Directions';
import SubwayIcon from '@mui/icons-material/Subway';
import ShieldIcon from '@mui/icons-material/Shield';
import Bolt from '@mui/icons-material/Bolt';
import AlternateEmailIcon from '@mui/icons-material/AlternateEmail';
import WidgetsIcon from '@mui/icons-material/Widgets';
import BrowserNotSupportedIcon from '@mui/icons-material/BrowserNotSupported';
import StarBorderIcon from '@mui/icons-material/StarBorder';
import RadioButtonCheckedIcon from '@mui/icons-material/RadioButtonChecked';
import { toast } from 'react-toastify';

import { Routes } from '../constants/Routes';
import categoryData from '../assets/data/FinnishCategories.json';
import { getRouterName, getTranslatedCategoryData, matchRouteWithCategory, parseRouterName } from '../utils/helpers';
import axios from 'axios';
import { get } from 'lodash';
import { useTranslation } from 'react-i18next';
import { getErrorMessage } from '../utils/errorUtil';
import HelpIcon from '@mui/icons-material/Help';

export interface IMenuItem {
  key: string;
  primary: string | JSX.Element;
  icon?: JSX.Element;
  to?: string;
  children?: IMenuItem[];
}

const fetchMaterialClass = (): IMenuItem[] => {
  const loadIcons = (category: string) => {
    switch (category) {
      case 'Kaaviot':
        return <SsidChartIcon />;
      case 'Liikennöinti':
        return <DirectionsIcon />;
      case 'Taitorakenteet':
        return <SubwayIcon />;
      case 'Turvalaitteet':
        return <ShieldIcon />;
      case 'Sähkörata':
        return <Bolt />;
      case 'Yhteystiedot':
        return <AlternateEmailIcon />;
      case 'Muut':
        return <WidgetsIcon />;
      case 'Käyttöohjeet':
        return <HelpIcon />;
      case 'Baliisisanomat':
        return <HelpIcon />;
      default:
        return <BrowserNotSupportedIcon />;
    }
  };
  return getTranslatedCategoryData(categoryData).map((data) => {
    return {
      key: data.category,
      primary: data.category,
      icon: loadIcons(data.category),
      children: data.subCategories.map((item) => {
        return {
          key: item,
          primary: item,
          to: `/${getRouterName(data.category)}/${getRouterName(item)}`,
        };
      }),
    };
  });
};

const fetchFavoriteCategories = async (): Promise<IMenuItem[]> => {
  try {
    const response = await axios.get('/api/database/favorites');
    const categories = await response.data;

    return categories
      ? categories.map((category: any) => {
          const rataextraRequestPage = get(category, 'categoryDataBase.rataextraRequestPage', '');
          return {
            key: category.id,
            primary: parseRouterName(rataextraRequestPage),
            icon: '',
            to: matchRouteWithCategory(Routes, rataextraRequestPage),
          };
        })
      : [];
  } catch (e) {
    throw e;
  }
};

const postFavoriteCategory = async (categoryRoute: string) => {
  try {
    const response = await axios.post(
      `/api/database/favorites`,
      { category: categoryRoute },
      { headers: { 'Content-Type': 'application/json' } },
    );
    const newCategory = await response.data;
    return newCategory;
  } catch (e: any) {
    toast(getErrorMessage(e), { type: 'error' });
    throw e;
  }
};

const deleteFavoriteCategory = async (categoryRoute: string) => {
  try {
    const response = await axios.delete(`/api/database/favorites?category=${categoryRoute}`);
    const newCategory = await response.data;
    return newCategory;
  } catch (e: any) {
    toast(getErrorMessage(e), { type: 'error' });
    throw e;
  }
};

export const MenuContext = React.createContext({
  menu: {},
  menuItems: [],
  favoriteCategories: [],
  resetMenu: () => {},
  menuHandler: (_: string) => {},
  addFavoriteHandler: (_: string) => {},
  removeFavoriteHandler: (_: string) => {},
  fileUploadDisabled: true,
  fileUploadDisabledHandler: (_: boolean) => {},
});

export const MenuContextProvider = (props: any) => {
  const { t } = useTranslation(['common']);
  const [favoriteCategories, setFavoriteCategories] = useState<IMenuItem[]>([]);
  const [fileUploadDisabled, setFileUploadDisabled] = useState(true);

  useEffect(() => {
    const fetchFavorites = async () => {
      const favorites = await fetchFavoriteCategories();
      setFavoriteCategories(favorites);
    };

    fetchFavorites();
  }, []);

  const menuItems: any = [
    {
      key: t('common:menu.frontpage'),
      primary: t('common:menu.frontpage'),
      icon: <InfoIcon />,
      to: Routes.HOME,
    },
    {
      key: t('common:menu.favorite'),
      primary: t('common:menu.favorite'),
      icon: <StarBorderIcon />,
      children: favoriteCategories,
    },
    ...fetchMaterialClass(),
    // Logout should always be the last menu item
    {
      key: t('common:menu.balise'),
      primary: t('common:menu.balise'),
      icon: <RadioButtonCheckedIcon />,
      to: Routes.BALISE,
    },
    {
      key: t('common:menu.logout'),
      primary: <Typography variant="subtitle2">{t('common:menu.logout')}</Typography>,
      icon: <LogoutIcon />,
      to: `${window.location.origin}/sso/logout?auth=1`,
    },
  ];

  const initMenu = () => {
    let menuData = {};
    menuItems.forEach((item: IMenuItem) => {
      menuData = {
        ...menuData,
        [item.key]: false,
      };
    });
    return menuData;
  };

  const [menu, setMenu] = useState<{ [key: string]: boolean }>(initMenu);

  const menuHandler = (key: string) => {
    setMenu({ ...menu, [key]: !menu[key] });
  };

  const resetMenu = () => {
    setMenu(initMenu);
  };

  const addFavoriteHandler = async (categoryRoute: string) => {
    try {
      await postFavoriteCategory(categoryRoute);
      // Refetch favorite categories after adding a new one
      const updatedFavorites = await fetchFavoriteCategories();
      setFavoriteCategories(updatedFavorites);
    } catch (e) {
      console.error('Error adding favorite:', e);
    }
  };

  const removeFavoriteHandler = async (categoryRoute: string) => {
    try {
      await deleteFavoriteCategory(categoryRoute);
      // Refetch favorite categories after removing one
      const updatedFavorites = await fetchFavoriteCategories();
      setFavoriteCategories(updatedFavorites);
    } catch (e) {
      console.error('Error removing favorite:', e);
    }
  };

  const fileUploadDisabledHandler = (isDisabled: boolean) => {
    setFileUploadDisabled(isDisabled);
  };

  return (
    <MenuContext.Provider
      value={{
        menu: menu,
        menuItems: menuItems,
        favoriteCategories: favoriteCategories as any,
        resetMenu: resetMenu,
        menuHandler: menuHandler,
        addFavoriteHandler: addFavoriteHandler,
        removeFavoriteHandler: removeFavoriteHandler,
        fileUploadDisabled: fileUploadDisabled,
        fileUploadDisabledHandler: fileUploadDisabledHandler,
      }}
    >
      {props.children}
    </MenuContext.Provider>
  );
};
