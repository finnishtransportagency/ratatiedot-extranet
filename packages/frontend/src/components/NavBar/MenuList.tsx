import { ExpandLess, ExpandMore } from '@mui/icons-material';
import { ListItem, ListItemButton, ListItemIcon, ListItemText } from '@mui/material';
import { Fragment, useContext } from 'react';
import { Colors } from '../../constants/Colors';
import { AppBarContext } from '../../contexts/AppBarContext';
import { IMenuItem, MenuContext } from '../../contexts/MenuContext';
import { useLocation } from 'react-router-dom';
import { capitalize } from 'lodash';
import { getRouterName } from '../../utils/helpers';
import { theme } from '../../styles/createTheme';

export const MenuList = () => {
  const { openDrawer, toggleDrawer } = useContext(AppBarContext);
  const { menu, menuItems, menuHandler } = useContext(MenuContext);

  const categoryHandler = (key: string) => {
    menuHandler(key);
    if (!openDrawer) {
      toggleDrawer();
    }
  };
  const { pathname } = useLocation();
  const routeName = getRouterName(pathname).split('/').filter(Boolean)[0];

  return (
    <>
      {menuItems.map((item: IMenuItem) => {
        const { key, primary, icon, to, children } = item;

        const selected =
          (key !== 'Logout' && capitalize(getRouterName(primary as string)) === capitalize(routeName)) ||
          (key === 'Landing' && pathname === '/');
        return (
          <Fragment key={key}>
            <ListItem disablePadding alignItems="flex-start" onClick={() => categoryHandler(key)}>
              <ListItemButton
                href={to ? to : ''}
                selected={selected}
                sx={{
                  '&.Mui-selected': {
                    backgroundColor: '#fff',
                    '&::before': {
                      content: '""',
                      position: 'absolute',
                      left: 0,
                      top: 0,
                      bottom: 0,
                      width: '4px',
                      backgroundColor: Colors.midblue,
                    },
                    '&.MuiListItemButton-root:hover': {
                      backgroundColor: theme.palette.action.hover,
                    },
                  },
                }}
              >
                <ListItemIcon>{icon}</ListItemIcon>
                <ListItemText
                  primary={primary}
                  sx={{
                    opacity: openDrawer ? 1 : 0,
                    '.MuiListItemText-primary': {
                      fontFamily: selected ? 'Exo2-Bold' : 'Exo2-Regular',
                    },
                  }}
                />
                {children && (menu[key as never] ? <ExpandLess /> : <ExpandMore />)}
              </ListItemButton>
            </ListItem>
            {children &&
              menu[key as never] &&
              children.map((child: IMenuItem) => {
                const selected = child.to === pathname;
                return (
                  <ListItemButton
                    sx={{
                      pl: 9,
                      whiteSpace: 'normal',
                      flexGrow: 'unset',
                      backgroundColor: Colors.lightgrey,
                      '&.Mui-selected': {
                        backgroundColor: Colors.midgrey,
                        '&.MuiListItemButton-root:hover': {
                          backgroundColor: theme.palette.action.hover,
                        },
                      },
                    }}
                    key={child.key}
                    href={child.to ? child.to : ''}
                    selected={child.to === pathname}
                  >
                    <ListItemText
                      primary={child.primary}
                      sx={{
                        '.MuiListItemText-primary': {
                          fontFamily: selected ? 'Exo2-Bold' : 'Exo2-Regular',
                        },
                      }}
                    />
                  </ListItemButton>
                );
              })}
          </Fragment>
        );
      })}
    </>
  );
};
