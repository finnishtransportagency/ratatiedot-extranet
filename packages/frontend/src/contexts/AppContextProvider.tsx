import { FC } from 'react';
import { SearchContextProvider } from './SearchContext';
import { AppBarContextProvider } from './AppBarContext';
import { MenuContextProvider } from './MenuContext';

const providers = [SearchContextProvider, AppBarContextProvider, MenuContextProvider];

const combineComponents = (...components: FC[]) => {
  return components.reduce(
    (AccumulatedComponents: any, CurrentComponent: any) => {
      return ({ children }: any): JSX.Element => {
        return (
          <AccumulatedComponents>
            <CurrentComponent>{children}</CurrentComponent>
          </AccumulatedComponents>
        );
      };
    },
    ({ children }: any) => <>{children}</>,
  );
};

export const AppContextProvider = combineComponents(...providers);
