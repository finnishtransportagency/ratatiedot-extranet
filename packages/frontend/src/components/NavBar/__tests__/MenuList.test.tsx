import { vi } from 'vitest';
import { render, screen } from '@testing-library/react';
import renderer, { act } from 'react-test-renderer';
import { MenuContextProvider } from '../../../contexts/MenuContext';
import { MenuList } from '../MenuList';

const mockedUsedNavigate = vi.fn();
vi.mock('react-router-dom', async () => {
  const actual = await vi.importActual('react-router-dom');
  return {
    ...(actual as Record<string, unknown>),
    useNavigate: () => mockedUsedNavigate,
  };
});

test('MenuList renders properly', () => {
  render(<MenuList />);
});

test('MenuList matches snapshot', () => {
  const menuList = renderer.create(<MenuList />).toJSON();
  expect(menuList).toMatchSnapshot();
});

test('MenuList, open a menu and check that new item is visible', async () => {
  // const view =
  render(
    <MenuContextProvider>
      <MenuList />
    </MenuContextProvider>,
  );
  // TODO: Replace closest with something better to fix linter error
  const charts = screen.getByText('Kaaviot').closest('li'); // eslint-disable-line testing-library/no-node-access
  act(() => {
    // Open Kaaviot
    charts?.click();
  });
  const lineCharts = screen.getByText('Linjakaaviot').closest('a'); // eslint-disable-line testing-library/no-node-access
  expect(lineCharts).toBeVisible();
  expect(lineCharts).toHaveAttribute('href', '/kaaviot/linjakaaviot');
});
