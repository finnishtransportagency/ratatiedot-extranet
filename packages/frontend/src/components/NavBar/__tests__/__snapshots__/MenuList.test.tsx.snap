// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuList matches snapshot 1`] = `
Array [
  <div>
    <li
      className="MuiListItem-root MuiListItem-gutters MuiListItem-alignItemsFlexStart css-dwve99-MuiListItem-root"
      disabled={false}
      onClick={[Function]}
    >
      <a
        className="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-16ac5r2-MuiButtonBase-root-MuiListItemButton-root"
        href="/"
        onBlur={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
      >
        <div
          className="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="InfoIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-6h2v6zm0-8h-2V7h2v2z"
            />
          </svg>
        </div>
        <div
          className="MuiListItemText-root css-1cag644-MuiListItemText-root"
        >
          <span
            className="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
          >
            Etusivu
          </span>
        </div>
      </a>
    </li>
  </div>,
  <div>
    <li
      className="MuiListItem-root MuiListItem-gutters MuiListItem-alignItemsFlexStart css-dwve99-MuiListItem-root"
      disabled={false}
      onClick={[Function]}
    >
      <div
        className="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-16ac5r2-MuiButtonBase-root-MuiListItemButton-root"
        href=""
        onBlur={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        role="button"
        tabIndex={0}
      >
        <div
          className="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="SsidChartIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M21 5.47 12 12 7.62 7.62 3 11V8.52L7.83 5l4.38 4.38L21 3v2.47zM21 15h-4.7l-4.17 3.34L6 12.41l-3 2.13V17l2.8-2 6.2 6 5-4h4v-2z"
            />
          </svg>
        </div>
        <div
          className="MuiListItemText-root css-1cag644-MuiListItemText-root"
        >
          <span
            className="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
          >
            Kaaviot
          </span>
        </div>
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="ExpandMoreIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
          />
        </svg>
      </div>
    </li>
  </div>,
  <div>
    <li
      className="MuiListItem-root MuiListItem-gutters MuiListItem-alignItemsFlexStart css-dwve99-MuiListItem-root"
      disabled={false}
      onClick={[Function]}
    >
      <div
        className="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-16ac5r2-MuiButtonBase-root-MuiListItemButton-root"
        href=""
        onBlur={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        role="button"
        tabIndex={0}
      >
        <div
          className="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="DirectionsIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="m21.41 10.59-7.99-8c-.78-.78-2.05-.78-2.83 0l-8.01 8c-.78.78-.78 2.05 0 2.83l8.01 8c.78.78 2.05.78 2.83 0l7.99-8c.79-.79.79-2.05 0-2.83zM13.5 14.5V12H10v3H8v-4c0-.55.45-1 1-1h4.5V7.5L17 11l-3.5 3.5z"
            />
          </svg>
        </div>
        <div
          className="MuiListItemText-root css-1cag644-MuiListItemText-root"
        >
          <span
            className="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
          >
            Liikenn√∂inti
          </span>
        </div>
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="ExpandMoreIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
          />
        </svg>
      </div>
    </li>
  </div>,
  <div>
    <li
      className="MuiListItem-root MuiListItem-gutters MuiListItem-alignItemsFlexStart css-dwve99-MuiListItem-root"
      disabled={false}
      onClick={[Function]}
    >
      <div
        className="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-16ac5r2-MuiButtonBase-root-MuiListItemButton-root"
        href=""
        onBlur={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        role="button"
        tabIndex={0}
      >
        <div
          className="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="SubwayIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <circle
              cx="15.5"
              cy="16"
              r="1"
            />
            <circle
              cx="8.5"
              cy="16"
              r="1"
            />
            <path
              d="M7.01 9h10v5h-10zM17.8 2.8C16 2.09 13.86 2 12 2c-1.86 0-4 .09-5.8.8C3.53 3.84 2 6.05 2 8.86V22h20V8.86c0-2.81-1.53-5.02-4.2-6.06zm.2 13.08c0 1.45-1.18 2.62-2.63 2.62l1.13 1.12V20H15l-1.5-1.5h-2.83L9.17 20H7.5v-.38l1.12-1.12C7.18 18.5 6 17.32 6 15.88V9c0-2.63 3-3 6-3 3.32 0 6 .38 6 3v6.88z"
            />
          </svg>
        </div>
        <div
          className="MuiListItemText-root css-1cag644-MuiListItemText-root"
        >
          <span
            className="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
          >
            Taitorakenteet
          </span>
        </div>
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="ExpandMoreIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
          />
        </svg>
      </div>
    </li>
  </div>,
  <div>
    <li
      className="MuiListItem-root MuiListItem-gutters MuiListItem-alignItemsFlexStart css-dwve99-MuiListItem-root"
      disabled={false}
      onClick={[Function]}
    >
      <div
        className="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-16ac5r2-MuiButtonBase-root-MuiListItemButton-root"
        href=""
        onBlur={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        role="button"
        tabIndex={0}
      >
        <div
          className="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="ShieldIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 1 3 5v6c0 5.55 3.84 10.74 9 12 5.16-1.26 9-6.45 9-12V5l-9-4z"
            />
          </svg>
        </div>
        <div
          className="MuiListItemText-root css-1cag644-MuiListItemText-root"
        >
          <span
            className="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
          >
            Turvalaitteet
          </span>
        </div>
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="ExpandMoreIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
          />
        </svg>
      </div>
    </li>
  </div>,
  <div>
    <li
      className="MuiListItem-root MuiListItem-gutters MuiListItem-alignItemsFlexStart css-dwve99-MuiListItem-root"
      disabled={false}
      onClick={[Function]}
    >
      <div
        className="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-16ac5r2-MuiButtonBase-root-MuiListItemButton-root"
        href=""
        onBlur={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        role="button"
        tabIndex={0}
      >
        <div
          className="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="AlternateEmailIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10h5v-2h-5c-4.34 0-8-3.66-8-8s3.66-8 8-8 8 3.66 8 8v1.43c0 .79-.71 1.57-1.5 1.57s-1.5-.78-1.5-1.57V12c0-2.76-2.24-5-5-5s-5 2.24-5 5 2.24 5 5 5c1.38 0 2.64-.56 3.54-1.47.65.89 1.77 1.47 2.96 1.47 1.97 0 3.5-1.6 3.5-3.57V12c0-5.52-4.48-10-10-10zm0 13c-1.66 0-3-1.34-3-3s1.34-3 3-3 3 1.34 3 3-1.34 3-3 3z"
            />
          </svg>
        </div>
        <div
          className="MuiListItemText-root css-1cag644-MuiListItemText-root"
        >
          <span
            className="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
          >
            Yhteystiedot
          </span>
        </div>
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="ExpandMoreIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
          />
        </svg>
      </div>
    </li>
  </div>,
  <div>
    <li
      className="MuiListItem-root MuiListItem-gutters MuiListItem-alignItemsFlexStart css-dwve99-MuiListItem-root"
      disabled={false}
      onClick={[Function]}
    >
      <div
        className="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-16ac5r2-MuiButtonBase-root-MuiListItemButton-root"
        href=""
        onBlur={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        role="button"
        tabIndex={0}
      >
        <div
          className="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="WidgetsIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M13 13v8h8v-8h-8zM3 21h8v-8H3v8zM3 3v8h8V3H3zm13.66-1.31L11 7.34 16.66 13l5.66-5.66-5.66-5.65z"
            />
          </svg>
        </div>
        <div
          className="MuiListItemText-root css-1cag644-MuiListItemText-root"
        >
          <span
            className="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
          >
            Muut
          </span>
        </div>
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="ExpandMoreIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
          />
        </svg>
      </div>
    </li>
  </div>,
  <div>
    <li
      className="MuiListItem-root MuiListItem-gutters MuiListItem-alignItemsFlexStart css-dwve99-MuiListItem-root"
      disabled={false}
      onClick={[Function]}
    >
      <a
        className="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-16ac5r2-MuiButtonBase-root-MuiListItemButton-root"
        href="http://localhost/sso/logout?auth=1"
        onBlur={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
      >
        <div
          className="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="LogoutIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="m17 7-1.41 1.41L18.17 11H8v2h10.17l-2.58 2.58L17 17l5-5zM4 5h8V3H4c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h8v-2H4V5z"
            />
          </svg>
        </div>
        <div
          className="MuiListItemText-root css-1cag644-MuiListItemText-root"
        >
          <h6
            className="MuiTypography-root MuiTypography-subtitle2 css-ati914-MuiTypography-root"
          >
            Kirjaudu ulos
          </h6>
        </div>
      </a>
    </li>
  </div>,
]
`;
